### PR TITLE
feat: Add functionality to build simulator from a YAML configuration file

### PR DESCRIPTION
**Pull request by lsetiawan**
## Overview

This PR introduces a function called `build_simulator` which can take a YAML configuration file to build simulator from that file. An example YAML configuration can be:

```yaml
cosmology: &cosmo
    name: cosmo
    kind: FlatLambdaCDM
    
lens: &lens
    name: lens
    kind: SIE
    params:
        z_l: 1.0
        x0: 0.0
        y0: 0.01
        q: 0.5
        phi: pi / 3.0
        b: 1.0
    init_kwargs:
        cosmology: *cosmo
        
source: &source
    name: source
    kind: Sersic
    params:
        x0: 0.01
        y0: -0.03
        q: 0.6
        phi: -pi / 4
        n: 2.0
        Re: 0.5
        Ie: 1.0

lnslt: &lnslt
    name: lenslight
    kind: Sersic
    params:
        x0: 0.0
        y0: 0.01
        q: 0.7
        phi: pi / 4
        n: 3.0
        Re: 0.7
        Ie: 1.0
    
simulator:
    name: minisim
    kind: Lens_Source
    params:
        z_s: 2.0
    init_kwargs:
        # Single lense
        lens: *lens
        source: *source
        lens_light: *lnslt
        pixelscale: 0.05
        pixels_x: 50
        psf:
            # Special kind of input
            # this will not get validated other than
            # the fact that it is a dictionary 
            # with "func" and "kwargs" keys
            func: caustics.utils.gaussian
            kwargs:
                pixelscale: 0.05
                nx: 11
                ny: 12
                sigma: 0.2
                upsample: 2
```

With the above yaml file a user can create a simulator called "minisim" with `caustics.build_simulator`. This functionality leverages [pydantic v2](https://docs.pydantic.dev/latest/) and typing quite heavily to give a lot more metadata regarding the arguments that are inputs for each of the Parametrized sub-class.

### CODE REVIEW

File: src/caustics/models/utils.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+                )
+                if not isinstance(v, expected_type):
+                    if isinstance(v, dict):
+                        if all(k for k in ["func", "kwargs"]):
+                            # Special case for the init_kwargs
+                            # this is to allow for creating tensor with some
+                            # caustics utils function, such as
+                            # `caustics.utils.gaussian`
+                            func = _import_func_or_class(v["func"])
+                            v = func(**v["kwargs"])  # type: ignore
```
Comment (lsetiawan) [inline src/caustics/models/utils.py:158]: Raise error if not an accepted dict!

File: src/caustics/models/registry.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+    _categories = OrderedDict(
+        [
+            ("cosmology", list(cosmology.keys())),
+            ("single_lenses", list(single_lenses.keys())),
+            ("multi_lenses", list(multi_lenses.keys())),
+            ("light", list(light.keys())),
+            ("simulators", list(simulators.keys())),
+        ]
+    )
```
Comment (lsetiawan) [inline src/caustics/models/registry.py:43]:
```suggestion
```

Comment (lsetiawan) [reply src/caustics/models/registry.py:43]: On a separate PR.

**Comment by lsetiawan** [comment]
Apologies for the large PR @ConnorStoneAstro! I just added the tests in here to make the codecov overlord happier 😄

**Review by ConnorStoneAstro — COMMENTED** [review]
This looks really great! Now you just need to get the tests working right?

File: src/caustics/models/api.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    Build a simulator from the configuration
+    """
+    simulators = setup_simulator_models()
+    Config = create_model(
+        "Config", __base__=StateConfig, simulator=(simulators, Field(...))
+    )
+
+    # Load the yaml config
+    yaml_bytes = from_file(config_path)
+    config_json = yaml.safe_load(yaml_bytes)
```
[the remark this answers is not in the document]
Comment (lsetiawan) [reply src/caustics/models/api.py:23]: It's a dict. I'll make it clear. Thanks for catching! 😄

**Review by ConnorStoneAstro — APPROVED** [review]
Spectacular! :) I also tried it out in a codespace and it worked :) Ready for launch!

**Comment by ConnorStoneAstro** [comment]
I still think a docs page showing what this can do will be really helpful, but that can/should be a separate PR

**Comment by lsetiawan** [comment]
> I still think a docs page showing what this can do will be really helpful, but that can/should be a separate PR

This is coming soon! 😄

**Pull request merged**